src/util/inc/tutil.h

```diff
@@ -110,6 +110,23 @@ extern "C" {
 
 #define POW2(x) ((x) * (x))
 
+#define MEMNCPY(dst, src, n)       \
+  do {                             \
+    assert((n) <= sizeof(dst));    \
+    strncpy((dst), (src), n);      \
+  } while (0)
+
+#define STRNCPY(dst, src, n)       \
+  do {                             \
+    assert((n) <= sizeof(dst));    \
```

Comment:
it is not correct to use `sizeof(dst)` here if `dst` is a pointer instead of an array.

Reply:
yes, I know what you mean. Because it's C, I have no decent way to distinguish whether it's pointer or array. what else would you suggest, if you happen to know?

thx!

Reply:
pls revise `STRNCPY` as below and remove `MEMNCPY`:
```c
#define STRNCPY(dst, src, n) \
do { \
   strncpy((dst), (src), (n)); \
   (dst)[(n) - 1] = 0; \
} while (0);
```

the idea behind is: we should try to avoid behavior changes when redefine a standard API.

because what we need here is a `strncpy` like function but could guarantee destination string is null terminated, the behavior change should be limit to "null terminated", otherwise, it will be very possible for the new function/macro to be misused, as developers think it is just a `strncpy`。

 `MEMNCPY` is similiar, developers will think it is a `memcpy`, but it is a `strncpy`.

Reply:
yes, i know the idea behind. but i still have some uncertainties as following:
just take `src/client/src/tscSql.c` as example:
line 89:  `strncpy(pObj->user, user, TSDB_USER_LEN);`
whereby `pObj->user` is defined in `src/client/inc/tsclient.h` as 
284 typedef struct STscObj {
...
289   char               user[TSDB_USER_LEN];
...
this seems also a bug, will user is TSDB_USER_LEN long or TSDB_USER_LEN-1 long?

and then, we take another example in `src/kit/taosdump/taosdump.c`
line 391:  `strncpy(pTableRecordInfo->tableRecord.name, (char*)row[TSDB_SHOW_TABLES_NAME_INDEX], fields[TSDB_SHOW_TABLES_NAME_INDEX].bytes);`
again: is ...bytes including null-terminator or not?

having these two concerns in mind, I chose those STRNCPY/MEMNCPY as a preliminary solution, with assert to check (in debug mode), and never overwrite dest buffer.

in order to make the coding phylosophy consistent, your idea shall be followed as suggested. but i am afraid of being less clear about the program logic under the hood,  submission as suggestion might fail the test (for those boundary test cases).

even if i chose to change user[TSDB_USER_LEN]; to user[TSDB_USER_LEN+1];
that might potentially break somewhere else? i have no idea.

thus, MEMNCPY/STRNCPY, as defined this way temporarily, it'll introduce code-break as less as possible.

as a compromise, shall we submit this way first, and then check boundary-failure test later, and make a final decision?

sorry for this lengthy response.

Reply:
btw, look at those failed submissions under this pr, that tells the story that how less familiar that i am with the logic underneath till now. :)

Reply:
I totally understand your feeling, because it is also my feeling.

generally, when define the max length of a string, the terminator should be included. but in current code base, sometimes the terminator is included while sometimes not, and this is even happen in one variable!

however, this is not a reason for us to break the behavior of standard functions, because it will make things worse.

if some test cases do fail, we just fix them, from a long term view, this way makes the code better.

Reply:
ok